overlord/snapstate/autorefresh.go

```diff
@@ -254,49 +256,49 @@ func (m *autoRefresh) Ensure() error {
 		logger.Debugf("Next refresh scheduled for %s.", m.nextRefresh.Format(time.RFC3339))
 	}
 
-	// should we hold back refreshes?
-	holdTime, err := m.EffectiveRefreshHold()
+	held, holdTime, err := m.isRefreshHeld(refreshSchedule)
 	if err != nil {
 		return err
 	}
-	if holdTime.After(now) {
-		return nil
-	}
-	if !holdTime.IsZero() {
-		// expired hold case
-		m.clearRefreshHold()
-		if m.nextRefresh.Before(holdTime) {
-			// next refresh is obsolete, compute the next one
-			delta := timeutil.Next(refreshSchedule, holdTime, maxPostponement)
-			now = time.Now()
-			m.nextRefresh = now.Add(delta)
-		}
-	}
 
 	// do refresh attempt (if needed)
-	if !m.nextRefresh.After(now) {
-		var can bool
-		can, err = m.canRefreshRespectingMetered(now, lastRefresh)
-		if err != nil {
-			return err
-		}
-		if !can {
-			// clear nextRefresh so that another refresh time is calculated
-			m.nextRefresh = time.Time{}
-			return nil
+	if !held {
+		// expired hold case
+		if !holdTime.IsZero() {
+			m.clearRefreshHold()
+			if m.nextRefresh.Before(holdTime) {
+				// next refresh is obsolete, compute the next one
+				delta := timeutil.Next(refreshSchedule, holdTime, maxPostponement)
+				now = time.Now()
+				m.nextRefresh = now.Add(delta)
+			}
 		}
 
-		// Check that we have reasonable delays between attempts.
-		// If the store is under stress we need to make sure we do not
-		// hammer it too often
-		if !m.lastRefreshAttempt.IsZero() && m.lastRefreshAttempt.Add(refreshRetryDelay).After(time.Now()) {
-			return nil
-		}
+		// refresh is also "held" if the next time is after now
+		if !m.nextRefresh.After(now) {
```

Comment:
Could we avoid the negation by saying:
```
if m.nextRefresh.Before(now) { ....
```

Reply:
hmm actually Go is a little silly like this because in many codepaths we have that `now` is exactly equal to `m.nextRefresh`, and `After()` is really `greater than or equal to` while `Before()` is just strictly `less than`, so if I change to `Before`, then I also need to make it 
```go
		if m.nextRefresh.Before(now) || m.nextRefresh.Equals(now) {
```

which is less readable to me, so I think using `!After()` and a small comment is better

Reply:
the explanation is a bit confusing,  After is just >  and Before is just < ,  and Equal is ==,  but yes, this means !After is not Before

Reply:
ah yes sorry I missed that After() is false for the same times, but the `!` makes that true, I will push up a better explanatory comment then

Reply:
done, hopefully this is more clear now